### PR TITLE
fix: migrate storage to instanceName-apiKey to isolate by instance

### PR DESCRIPTION
**Pull request by justin-fiedler**
### Summary

* Storage prefix now includes both **apiKey** and **instanceName** to prevent unintended data sharing
* Added migration to migrate legacy storage to new format
* Only migrate events in sandboxed apps to prevent data pollution across orgs
* Always migrate UserDefaults (deviceId, userId, etc) - these seem to be always sandboxed
* Added STORAGE_VERSION to UserDefaults to prevent duplicate migrations
* Added tests

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-SDK-Template/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

### CODE REVIEW

File: Sources/Amplitude/Amplitude.swift [hunk 1]
```diff
@@ -41,10 +41,10 @@ public class Amplitude {
 
         migrateApiKeyStorages()
         migrateDefaultInstanceStorages()
-
-        if configuration.migrateLegacyData {
+        if configuration.migrateLegacyData && getStorageVersion() < .API_KEY_AND_INSTANCE_NAME {
             RemnantDataMigration(self).execute()
         }
+        migrateInstanceOnlyStorages()
```
Comment (Mercy811) [inline Sources/Amplitude/Amplitude.swift:47]: should we migrate instances only storages when current storage version is less than API_KEY_AND_INSTANCE_NAME?

Comment (justin-fiedler) [reply Sources/Amplitude/Amplitude.swift:47]: Yep, there is a check is in the `migrateInstanceOnlyStorages ()` method

File: Sources/Amplitude/Types.swift [hunk 2]
```diff
@@ -74,6 +74,23 @@ public enum StorageKey: String, CaseIterable {
     case DEVICE_ID = "device_id"
     case APP_BUILD = "app_build"
     case APP_VERSION = "app_version"
+    // The version of PersistentStorage, used for data migrations
+    // Value should be a PersistentStorageVersion value
+    // Note the first version is 2, which corresponds to instanceName-apiKey based storage
```
Comment (Mercy811) [inline Sources/Amplitude/Types.swift:79]: nit: 
```suggestion
    // Note the first version is 2, which corresponds toapiKey-instanceName based storage
```

File: Tests/AmplitudeTests/AmplitudeTests.swift [hunk 3]
```diff
@@ -324,6 +324,171 @@ final class AmplitudeTests: XCTestCase {
...
+            flushIntervalMillis: 99999,
+            logLevel: LogLevelEnum.DEBUG,
+            defaultTracking: DefaultTrackingOptions.NONE
+        )
+
+        // Create storages using instance name only
+        let legacyEventStorage = PersistentStorage(storagePrefix: "storage-\(config.getNormalizeInstanceName())")
+        let legacyIdentityStorage = PersistentStorage(storagePrefix: "identify-\(config.getNormalizeInstanceName())")
+
+        print("Creating legacy Amplitude instance")
```
Comment (Mercy811) [inline Tests/AmplitudeTests/AmplitudeTests.swift:342]: nit: is it left intentional? Not sure if printing to console is a good practices in swift. 

The same for `print(legacyEventsString)` and `print(eventsString)` below.

**Review by Mercy811 — APPROVED** [review]
Thanks Justin for fixing this. This is huge on security.

File: Tests/AmplitudeTests/AmplitudeTests.swift (hunk 3 above)
Comment (justin-fiedler) [reply Tests/AmplitudeTests/AmplitudeTests.swift]: 😅

File: Sources/Amplitude/Types.swift (hunk 2 above)
Comment (justin-fiedler) [reply Sources/Amplitude/Types.swift]: Fixed

File: Tests/AmplitudeTests/AmplitudeTests.swift (hunk 3 above)
Comment (justin-fiedler) [reply Tests/AmplitudeTests/AmplitudeTests.swift]: Fixed

**Review by izaaz — APPROVED** [review]

**Pull request merged**